router/data_store.go

```diff
@@ -89,38 +98,224 @@ INSERT INTO ` + tableNameTCP + ` (parent_ref, service, leader, port)
+			err = ErrInvalid
+		}
+		return err
 	}
-	return err
+	return nil
+}
+
+func (d *pgDataStore) addHTTP(r *router.Route) error {
+	tx, err := d.pgx.Begin()
+	if err != nil {
+		return err
+	}
+	if err := tx.QueryRow(
+		sqlAddRouteHTTP,
+		r.ParentRef,
+		r.Service,
+		r.Leader,
+		r.Domain,
+		r.Sticky,
+		r.Path,
+	).Scan(&r.ID, &r.CreatedAt, &r.UpdatedAt); err != nil {
+		tx.Rollback()
+		return err
+	}
+	if err := d.addRouteCertWithTx(tx, r); err != nil {
+		tx.Rollback()
+		return err
+	}
+	return tx.Commit()
+}
+
+func (d *pgDataStore) addTCP(r *router.Route) error {
+	return d.pgx.QueryRow(
+		sqlAddRouteTCP,
+		r.ParentRef,
+		r.Service,
+		r.Leader,
+		r.Port,
+	).Scan(&r.ID, &r.CreatedAt, &r.UpdatedAt)
+}
+
+const sqlSelectCert = `
+SELECT id, created_at, updated_at FROM ` + tableNameCertificates + `
+	WHERE tls_cert_sha256 = $1 AND deleted_at IS NULL`
+
+const sqlAddCert = `
+INSERT INTO ` + tableNameCertificates + ` (tls_cert, tls_key, tls_cert_sha256)
+	VALUES ($1, $2, $3)
+	RETURNING id, created_at, updated_at
+`
+
+const sqlAddRouteCert = `
+INSERT INTO ` + tableNameRoutesCertificate + ` (http_route_id, certificate_id)
+	VALUES ($1, $2)
+`
+
+const sqlCleanupRouteCerts = `
+DELETE FROM ` + tableNameRoutesCertificate + `
+	WHERE http_route_id = $1`
+
+func (d *pgDataStore) AddCert(c *router.RouteCert) error {
+	tx, err := d.pgx.Begin()
+	if err != nil {
+		return err
+	}
+	if err := d.addCertWithTx(tx, c); err != nil {
+		tx.Rollback()
+		return err
+	}
+	return tx.Commit()
+}
+
+func (d *pgDataStore) addCertWithTx(tx *pgx.Tx, c *router.RouteCert) error {
+	tlsCertSHA256 := hex.EncodeToString(sha256.New().Sum([]byte(c.TLSCert)))
+	if err := tx.QueryRow(sqlSelectCert, tlsCertSHA256).Scan(&c.ID, &c.CreatedAt, &c.UpdatedAt); err != nil {
+		if err := tx.QueryRow(sqlAddCert, c.TLSCert, c.TLSKey, tlsCertSHA256).Scan(&c.ID, &c.CreatedAt, &c.UpdatedAt); err != nil {
+			return err
+		}
+	}
+	for _, rid := range c.Routes {
+		if _, err := tx.Exec(sqlCleanupRouteCerts, rid); err != nil {
+			return err
+		}
+		if _, err := tx.Exec(sqlAddRouteCert, rid, c.ID); err != nil && !postgres.IsUniquenessError(err, "") {
+			return err
+		}
+	}
+	return nil
+}
+
+func (d *pgDataStore) addRouteCertWithTx(tx *pgx.Tx, r *router.Route) error {
+	var cert *router.RouteCert
+	if r.LegacyTLSCert != "" || r.LegacyTLSKey != "" {
+		cert = &router.RouteCert{
+			TLSCert: r.LegacyTLSCert,
+			TLSKey:  r.LegacyTLSKey,
+		}
+	} else {
+		cert = r.Certificate
+	}
+	if cert == nil {
+		return nil
+	}
+	cert.Routes = []string{r.ID}
+	if err := d.addCertWithTx(tx, cert); err != nil {
+		return err
+	}
+	r.Certificate = &router.RouteCert{
+		ID:        cert.ID,
+		TLSCert:   cert.TLSCert,
+		TLSKey:    cert.TLSKey,
+		CreatedAt: cert.CreatedAt,
+		UpdatedAt: cert.UpdatedAt,
+	}
+	return nil
+}
+
+const sqlGetCert = `
+SELECT ` + selectColumnsHTTPCert + `, ARRAY(
+	SELECT http_route_id::varchar FROM ` + tableNameRoutesCertificate + ` WHERE certificate_id = $1
```

Comment:
Is it necessary to cast to varchar here? I don't think it makes much difference either way just wondering. If so then there is a problem in the array handling of pgx.